File: data-structure-and-algorithm/java/src/com/bestlyg/set/TreeSet2.java
package com.bestlyg.set;

import java.util.Comparator;

import com.bestlyg.tree.BinaryTree;
import com.bestlyg.tree.RBTree;

public class TreeSet2<E> implements Set<E> {
	private RBTree<E> tree;

	public TreeSet2() {
		this(null);
	}

	public TreeSet2(Comparator<E> comparator) {
		tree = new RBTree<>(comparator);
	}

	@Override
	public int size() {
		return tree.size();
	}

	@Override
	public boolean isEmpty() {
		return tree.isEmpty();
	}

	@Override
	public void clear() {
		tree.clear();
	}

	@Override
	public boolean contains(E element) {
		return tree.contains(element);
	}

	@Override
	public void add(E element) {
		tree.add(element);
	}

	@Override
	public void remove(E element) {
		tree.remove(element);
	}

	@Override
	public void traversal(Visitor<E> visitor) {
		tree.inorder(new BinaryTree.Visitor<E>() {
			@Override
			public boolean visit(E element) {
				return visitor.visit(element);
			}
		});
	}

}

File: data-structure-and-algorithm/typescript/src/core/set/TreeSet2.ts
import RedBlackTree from "../tree/RedBlackTree";
import ISet from "./ISet";
import { Comparator } from "../../types";
/**
 * 内置红黑树
 */
export default class TreeSet2<T> implements ISet<T> {
  private _tree: RedBlackTree<T>;
  constructor(comparator: Comparator<T>) {
    this._tree = new RedBlackTree<T>(comparator);
  }
  size(): number {
    return this._tree.size();
  }
  isEmpty(): boolean {
    return this._tree.isEmpty();
  }
  clear(): void {
    this._tree.clear();
  }
  contains(element: T): boolean {
    return this._tree.contains(element);
  }
  add(element: T): void {
    this._tree.add(element);
  }
  remove(element: T): void {
    this._tree.remove(element);
  }
  traversal(visitor: (element: T) => boolean) {
    this._tree.inorder(visitor);
  }
}
